### PR TITLE
Functor documentation and a conversion utility

### PR DESCRIPTION
**Pull request by GiudGiud**

### CODE REVIEW

File: framework/doc/content/source/materials/MaterialFunctorConverter.md [hunk 1]
```diff
@@ -0,0 +1,10 @@
+# MaterialFunctorConverter
+
+The `MaterialFunctorConverter` is used to explicitly convert functors, such as [functor material properties](syntax/FunctorMaterials/index.md)
+into regular nonAD or AD material properties.
```
Comment (lindsayad) [inline framework/doc/content/source/materials/MaterialFunctorConverter.md:4]:
```suggestion
into regular non-AD or AD material properties.
```

File: framework/doc/content/syntax/Materials/index.md [hunk 2]
```diff
@@ -222,142 +222,8 @@ error.
...
-order to achieve a perfect Jacobian. Not surprisingly, if a functor evaluation is cached, then
-memory usage will increase.
-
-!alert note title=Caching Implementations
-Functor caching is only currently implemented for `ElemQpArg` and `ElemSideQpArg` spatial
-overloads. This is with the idea that calls to `FE::reinit` can be fairly expensive whereas for the
-other spatial argument types, evaluation of the functor (material property) may be relatively
-inexpensive compared to the memory expense incurred from caching. We may definitely implement
-caching for other overloads, however, if use cases call for it.
+Functor materials are special kind of materials used for on-the-fly material property evaluation.
```
Comment (lindsayad) [inline framework/doc/content/syntax/Materials/index.md:225]:
```suggestion
Functor materials are a special kind of materials used for on-the-fly material property evaluation.
```

File: framework/doc/content/syntax/FunctorMaterials/index.md [hunk 3]
```diff
@@ -0,0 +1,169 @@
+# FunctorMaterials
+
+Functor material properties are properties that are evaluated on-the-fly.
+Along with numerous classes of MOOSE objects, they are [Functors](syntax/Functors/index.md). This allows
```
Comment (lindsayad) [inline framework/doc/content/syntax/FunctorMaterials/index.md:4]:
```suggestion
Along with multiple other classes of MOOSE objects, they are [Functors](syntax/Functors/index.md). This allows
```
Only three main systems are functors at the current time, so numerous feels like a bit too much. Maybe not forever though!

File: framework/doc/content/syntax/FunctorMaterials/index.md [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+for the creation of objects with considerable inter-operability between systems.
+
+## Using functor materials
+
+Much like regular materials declare regular material properties, functor materials declare functor material properties.
+Following a producer/consumer model, `FunctorMaterials` produce functor material properties, which kernels, boundary conditions, other
+functor materials, etc, may consume.
+Both regular and functor material properties support a wide array of data types (scalar, vector, tensors,
+[automatic differentiation](automatic_differentiation/index.md) versions of all of these). Functor materials, because they are evaluated on-the-fly
+and call other functors on-the-fly as well, do not require dependency ordering.
```
Comment (lindsayad) [inline framework/doc/content/syntax/FunctorMaterials/index.md:14]: Yes, another very important benefit which I often forget to bring up. Thanks for highlighting this!

File: framework/doc/content/syntax/FunctorMaterials/index.md [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+All functor materials support caching in some capacity. This can avoid expensive material property computations, but is
+disabled by default due to the potential memory cost. The reader is referred to the [Functors caching documentation](syntax/Functors/index.md#caching)
+
+!alert warning
+Functor materials are generally **NOT** compatible with regular materials, as regular material properties **CANNOT** be used
+in lieu of functor material properties. Functor material properties can only be used as regular material properties using a
+[MaterialFunctorConverter.md] to process them.
+
+Functor materials can be created within the `[FunctorMaterials]` block, and currently the `[Materials]` block as well,
+but that syntax is considered for deprecation.
```
Comment (lindsayad) [inline framework/doc/content/syntax/FunctorMaterials/index.md:25]:
```suggestion
but that syntax will be deprecated.
```

File: framework/doc/content/syntax/FunctorMaterials/index.md [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+you may use the [!param](/Debug/SetupDebugAction/show_functors) to get more information about what functors were
+created and requested.
+
+## Developing with functor materials
+
+Functor material properties are properties that are evaluated
+on-the-fly. E.g. they can be viewed as functions of the current location in
+space (and time). Functor material properties provide several overloads of the
+`operator()` method for different "geometric quantities". One example of a
+"geometric quantity" is a `const Elem *`, e.g. for an `FVElementalKernel`, the
```
Comment (lindsayad) [inline framework/doc/content/syntax/FunctorMaterials/index.md:38]:
```suggestion
"geometric quantity" is based around an element, e.g. for an `FVElementalKernel`, the
```

File: framework/doc/content/syntax/FunctorMaterials/index.md [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+value of a functor material property in a cell-averaged sense can be obtained by
+the syntax
+
+- `_foo(_current_elem)`
```
Comment (lindsayad) [inline framework/doc/content/syntax/FunctorMaterials/index.md:42]:
```suggestion
- `_foo(makeElemArg(_current_elem), determineState())`
```

File: framework/doc/content/syntax/FunctorMaterials/index.md [hunk 8]
```diff
@@ -0,0 +1,169 @@
...
+
+where here `_foo` is a functor material property data member of the kernel. The
```
Comment (lindsayad) [inline framework/doc/content/syntax/FunctorMaterials/index.md:44]:
```suggestion
where here `_foo` is a functor data member of the kernel, `makeElemArg` is a helper routine for creating a functor element-based spatial argument, and `determineState()` is a helper routine for determining the correct time state to evaluate at, e.g. the current time for an implicit kernel and the old time for an explicit kernel. The
```

File: framework/doc/content/syntax/FunctorMaterials/index.md [hunk 9]
```diff
@@ -0,0 +1,169 @@
...
+
+Currently, functor material property evaluations are defined using the API:
+
+```c++
+template <typename T>
+template <typename PolymorphicLambda>
+void FunctorMaterialProperty<T>::
+setFunctor(const MooseMesh & mesh,
+           const std::set<SubdomainID> & block_ids,
+           PolymorphicLambda my_lammy);
```
Comment (lindsayad) [inline framework/doc/content/syntax/FunctorMaterials/index.md:68]:
```suggestion
template <typename T, typename PolymorphicLambda>
const Moose::FunctorBase<T> &
FunctorMaterial::addFunctorProperty(const std::string & name,
                                    PolymorphicLambda my_lammy,
                                    const std::set<ExecFlagType> & clearance_schedule = {
                                        EXEC_ALWAYS});
```

File: framework/doc/content/syntax/FunctorMaterials/index.md [hunk 10]
```diff
@@ -0,0 +1,169 @@
...
+```
+
+where the first two arguments are used to setup block restriction and the last argument is a lambda
+defining the property evaluation. The lambda must be callable with two arguments, the first
```
Comment (lindsayad) [inline framework/doc/content/syntax/FunctorMaterials/index.md:72]:
```suggestion
where the first argument will be the functor name stored in the problem functor database (all functor names must be unique), the second argument is a lambda
defining the property evaluation, and the third optional argument is a set defining at which execution stages the functor evaluation cache should be cleared. The lambda must be callable with two arguments, the first
```

**Review by lindsayad — CHANGES_REQUESTED** [review]

**Comment by lindsayad** [comment]
Can you let me know when you're done pushing commits/when this is ready for review? I was in the middle of reviewing and then I was told I needed to refresh because of your force push.

**Comment by GiudGiud** [comment]
oh force push was just for the design file in the tests. No change anywhere else
my b

File: framework/doc/content/syntax/FunctorMaterials/index.md [hunk 11]
```diff
@@ -0,0 +1,169 @@
...
+corresponding to space, and the second corresponding to time, and must return the type `T` of the
+`FunctorMaterialProperty`. An example of setting a constant functor material property that returns
+an `ADReal` looks like:
+
+```c++
+    _constant_unity_prop.setFunctor(
```
Comment (lindsayad) [inline framework/doc/content/syntax/FunctorMaterials/index.md:78]:
```suggestion
    addFunctorProperty<Real>(
```

File: framework/doc/content/syntax/FunctorMaterials/index.md [hunk 12]
```diff
@@ -0,0 +1,169 @@
...
+setFunctor(const MooseMesh & mesh,
+           const std::set<SubdomainID> & block_ids,
+           PolymorphicLambda my_lammy);
+```
+
+where the first two arguments are used to setup block restriction and the last argument is a lambda
+defining the property evaluation. The lambda must be callable with two arguments, the first
+corresponding to space, and the second corresponding to time, and must return the type `T` of the
+`FunctorMaterialProperty`. An example of setting a constant functor material property that returns
+an `ADReal` looks like:
```
Comment (lindsayad) [inline framework/doc/content/syntax/FunctorMaterials/index.md:75]:
```suggestion
a `Real` looks like:
```

File: framework/doc/content/syntax/FunctorMaterials/index.md [hunk 13]
```diff
@@ -0,0 +1,169 @@
...
+
+```c++
+    _constant_unity_prop.setFunctor(
+        _mesh, blockIDs(), [](const auto &, const auto &) -> ADReal { return 1.; });
```
Comment (lindsayad) [inline framework/doc/content/syntax/FunctorMaterials/index.md:79]:
```suggestion
        "foo", [](const auto &, const auto &) -> Real { return 1.; });
```

File: framework/doc/content/syntax/FunctorMaterials/index.md [hunk 14]
```diff
@@ -0,0 +1,169 @@
...
+```
+
+An example of a functor material property that depends on a nonlinear variable would look like
+
+```c++
+    _u_prop.setFunctor(_mesh, blockIDs(), [this](const auto & r, const auto & t) -> ADReal {
+      return _u_var(r, t);
+    });
```
Comment (lindsayad) [inline framework/doc/content/syntax/FunctorMaterials/index.md:87]:
```suggestion
  addFunctorProperty<ADReal>(_density_name,
                             [this](const auto & r, const auto & t) -> ADReal
                             { return _fluid.rho_from_p_T(_pressure(r, t), _temperature(r, t)); });
```

File: framework/doc/content/syntax/FunctorMaterials/index.md [hunk 15]
```diff
@@ -0,0 +1,169 @@
...
+```
+
+In the above example, we simply forward the calling arguments along to the variable. Variable
```
Comment (lindsayad) [inline framework/doc/content/syntax/FunctorMaterials/index.md:90]:
```suggestion
In the above example, we simply forward the calling arguments along to the captured functors. Variable
```

File: framework/doc/content/syntax/FunctorMaterials/index.md [hunk 16]
```diff
@@ -0,0 +1,169 @@
...
+space and time, e.g. the same lambda defining the property evaluation should apply across all
+spatial and temporal arguments. However, in the case that a functor material developer wishes to
+create specific implementations for specific arguments (as illustrated in `IMakeMyOwnFunctorProps`
+test class) or simply wishes to know more about the system, we give the details below.
+
+Any call to a functor (material property) looks like the following
+`_foo(const SpatialArg & r, const TemporalArg & t)`. Below are the possible type overloads of
+`SpatialArg`.
+
+### FaceArg id=spatial-overloads
```
Comment (lindsayad) [inline framework/doc/content/syntax/FunctorMaterials/index.md:108]: Maybe we should move all of this to the new Functors `index.md`. The following will also need to be updated

Comment (lindsayad) [reply framework/doc/content/syntax/FunctorMaterials/index.md:108]: And I'm happy to help with the updates

File: framework/doc/content/syntax/FunctorMaterials/index.md [hunk 17]
```diff
@@ -0,0 +1,169 @@
...
+corresponding to space, and the second corresponding to time, and must return the type `T` of the
+`FunctorMaterialProperty`. An example of setting a constant functor material property that returns
+an `ADReal` looks like:
+
+```c++
+    _constant_unity_prop.setFunctor(
+        _mesh, blockIDs(), [](const auto &, const auto &) -> ADReal { return 1.; });
+```
+
+An example of a functor material property that depends on a nonlinear variable would look like
```
Comment (lindsayad) [inline framework/doc/content/syntax/FunctorMaterials/index.md:82]:
```suggestion
An example of a functor material property that depends on a fluid properties user object and pressure and temperature functors looks like
```

File: framework/doc/content/syntax/FunctorMaterials/index.md (hunk 16 above)
Comment (lindsayad) [reply framework/doc/content/syntax/FunctorMaterials/index.md:108]: Oh it looks like we already do have it there. So let's remove this duplicate stuff

File: framework/doc/content/syntax/Functors/index.md [hunk 18]
```diff
@@ -0,0 +1,94 @@
+# Functor system
+
+Functors are an abstraction, existing as a base class, that is available to numerous systems in MOOSE:
```
Comment (lindsayad) [inline framework/doc/content/syntax/Functors/index.md:3]:
```suggestion
Functors are an abstraction, existing as a base class, that is available to several systems in MOOSE:
```

File: framework/doc/content/syntax/Functors/index.md [hunk 19]
```diff
@@ -0,0 +1,94 @@
...
+space and time, e.g. the same lambda defining the property evaluation should apply across all
+spatial and temporal arguments. However, in the case that a functor developer wishes to
+create specific implementations for specific arguments (as illustrated in `IMakeMyOwnFunctorProps`
+test class) or simply wishes to know more about the system, we give the details below.
+
+Any call to a functor looks like the following
+`_foo(const SpatialArg & r, const TemporalArg & t)`. Below are the possible type overloads of
+`SpatialArg`.
+
+### FaceArg id=spatial-overloads
```
Comment (lindsayad) [inline framework/doc/content/syntax/Functors/index.md:34]: Need to update all of these. I can make a pull request against your branch if you want some help sharing the load here

File: framework/doc/content/syntax/Functors/index.md [hunk 20]
```diff
@@ -0,0 +1,94 @@
...
+
+If a functor is a function of nonlinear degrees of freedom, evaluation with this
+argument will likely result in calls to libMesh `FE::reinit`.
+
+## Functor caching id=caching
+
+By default, functors are always (re-)evaluated every time they are called with
+`operator()`. However, the base class of functors, `Moose::Functor`, has a
+`setCacheClearanceSchedule(const std::set<ExecFlagType> & clearance_schedule)` API that allows
+control of evaluations. Supported values for the `clearance_schedule` are any combination of
```
Comment (lindsayad) [inline framework/doc/content/syntax/Functors/index.md:80]:
```suggestion
control of evaluations (in addition to the `clearance_schedule` argument to `addFunctorProperty` introduced above). Supported values for the `clearance_schedule` are any combination of
```

File: framework/doc/content/syntax/Functors/index.md [hunk 21]
```diff
@@ -0,0 +1,94 @@
...
+  evaluation of the ith point
+- The quadrature rule that can be used to initialize the functor on the given element and side
+
+If a functor is a function of nonlinear degrees of freedom, evaluation with this
+argument will likely result in calls to libMesh `FE::reinit`.
+
+## Functor caching id=caching
+
+By default, functors are always (re-)evaluated every time they are called with
+`operator()`. However, the base class of functors, `Moose::Functor`, has a
```
Comment (lindsayad) [inline framework/doc/content/syntax/Functors/index.md:78]:
```suggestion
`operator()`. However, the base class of functors, `Moose::FunctorBase`, has a
```

File: framework/include/materials/MaterialFunctorConverter.h [hunk 22]
```diff
@@ -0,0 +1,41 @@
...
+  MaterialFunctorConverterTempl(const InputParameters & parameters);
+
+protected:
+  virtual void computeQpProperties() override;
+  void initQpStatefulProperties() override;
+
+  /// Number of material properties to create
+  const std::size_t _num_functors_to_convert;
+
+  /// Incoming functors to convert
```
Comment (lindsayad) [inline framework/include/materials/MaterialFunctorConverter.h:32]:
```suggestion
  /// Incoming functors to convert. We up-convert non-AD functors to AD functors to ease implementation
```

File: framework/src/materials/MaterialFunctorConverter.C [hunk 23]
```diff
@@ -0,0 +1,91 @@
...
+
+registerMooseObject("MooseApp", MaterialFunctorConverter);
+registerMooseObject("MooseApp", VectorMaterialFunctorConverter);
+
+template <typename T>
+InputParameters
+MaterialFunctorConverterTempl<T>::validParams()
+{
+  InputParameters params = Material::validParams();
+  params.addClassDescription("Converts functor to nonAD and AD regular material properties");
```
Comment (lindsayad) [inline framework/src/materials/MaterialFunctorConverter.C:22]:
```suggestion
  params.addClassDescription("Converts functor to non-AD and AD regular material properties");
```

File: framework/src/materials/MaterialFunctorConverter.C [hunk 24]
```diff
@@ -0,0 +1,91 @@
...
+}
+
+template <typename T>
+MaterialFunctorConverterTempl<T>::MaterialFunctorConverterTempl(const InputParameters & parameters)
+  : Material(parameters),
+    _num_functors_to_convert(getParam<std::vector<MooseFunctorName>>("functors_in").size())
+{
+  auto functors_in = getParam<std::vector<MooseFunctorName>>("functors_in");
+  auto reg_props_out = getParam<std::vector<MaterialPropertyName>>("reg_props_out");
+  auto ad_props_out = getParam<std::vector<MaterialPropertyName>>("ad_props_out");
```
Comment (lindsayad) [inline framework/src/materials/MaterialFunctorConverter.C:39]:
```suggestion
  const auto & functors_in = getParam<std::vector<MooseFunctorName>>("functors_in");
  const auto & reg_props_out = getParam<std::vector<MaterialPropertyName>>("reg_props_out");
  const auto & ad_props_out = getParam<std::vector<MaterialPropertyName>>("ad_props_out");
```

File: framework/src/materials/MaterialFunctorConverter.C [hunk 25]
```diff
@@ -0,0 +1,91 @@
...
+{
+  computeQpProperties();
+}
+
+template <typename T>
+void
+MaterialFunctorConverterTempl<T>::computeQpProperties()
+{
+  // TODO: do we know when we are on a face or an element?
+  const Moose::ElemQpArg arg = {_current_elem, _qp, _qrule};
```
Comment (lindsayad) [inline framework/src/materials/MaterialFunctorConverter.C:79]: We should check `_bnd`. If `true`, then we need to make an `ElemSideQpArg` instead

File: test/tests/materials/functor_conversion/tests [hunk 26]
```diff
@@ -0,0 +1,40 @@
...
+      cli_args = "Materials/convert_to_reg/ad_props_out='m1 m2 m3 m4'"
+      expect_err = 'We dont support converting functors to both AD and regular material properties'
+      detail = 'if the user requests conversions to both AD and regular material properties, as this is not supported,'
+    []
+    [wrong_size]
+      type = RunException
+      input = conversion.i
+      cli_args = "Materials/convert_to_reg/reg_props_out='m1 m2 m3'"
+      expect_err = 'The number of output properties must match the number of input functors'
+      detail = 'if the user inputs a different number of functors and material properties, as the conversion match one to one, and'
```
Comment (lindsayad) [inline test/tests/materials/functor_conversion/tests:37]: I expect another detail with this terminating `and`

**Review by lindsayad — CHANGES_REQUESTED** [review]

**Comment by lindsayad** [comment]
Ok done with this round of review!

File: framework/doc/content/syntax/Functors/index.md (hunk 19 above)
Comment (GiudGiud) [reply framework/doc/content/syntax/Functors/index.md:34]: can do it

File: test/tests/materials/functor_conversion/tests (hunk 26 above)
Comment (GiudGiud) [inline test/tests/materials/functor_conversion/tests:37]:
```suggestion
      detail = 'if the user inputs a different number of functors and material properties, as the conversion match one to one.'
```

File: framework/src/materials/MaterialFunctorConverter.C (hunk 25 above)
Comment (GiudGiud) [reply framework/src/materials/MaterialFunctorConverter.C]: done

**Comment by GiudGiud** [comment]
Should be better now.
My apologies for not reading through the massive copy paste of outdated docs

File: framework/doc/content/syntax/Functors/index.md [hunk 27]
```diff
@@ -0,0 +1,138 @@
...
+## Developing with functors
+
+Functors are evaluated on-the-fly. E.g. they can be viewed as functions of the current location in
+space (and time). Functors provide several overloads of the
+`operator()` method for different "geometric quantities". One example of a
+"geometric quantity" is based around an element, e.g. for an `FVElementalKernel`, the
+value of a functor material property in a cell-averaged sense can be obtained by
+the syntax
+
+- `_foo(makeElemArg(_current_elem), determineState())`
```
Comment (joshuahansel) [inline framework/doc/content/syntax/Functors/index.md:33]: Before this, you should talk about declaring the member variable (and what type).

**Review by joshuahansel — CHANGES_REQUESTED** [review]
I still need to review more - I just wanted to take a break. I'm just reviewing the doc by the way.

File: framework/doc/content/syntax/Functions/index.md [hunk 28]
```diff
@@ -37,6 +37,20 @@ Check out `PiecewiseBilinear` to see how to update a function to support AD by
...
+objects that can compute values at a location in space and time.
+
+As `Functors`, they may be specified to objects such as the
+[FunctorElementalAux.md] in their functor argument. This vastly expands the number
+of objects that can use `Functions` to compute spatial quantities.
+
+!alert note
+When making a new object using `Functions` to contribute back to MOOSE,
+we ask that you consider using [Functors](syntax/Functors/index.md) instead
+to naturally enable its use with variables and functor material properties.
```
Comment (lindsayad) [inline framework/doc/content/syntax/Functions/index.md:52]: Very nice note

File: framework/doc/content/syntax/FunctorMaterials/index.md [hunk 29]
```diff
@@ -0,0 +1,108 @@
...
+where `TYPE` can be anything such as `Real, ADReal, RealVectorValue, ADRealVectorValue`
+etc. To get a functor material property:
+
+- `getFunctor<TYPE>`
+
+It's worth noting that whereas the traditional regular material property system
+has different methods to declare/get non-AD and AD properties, the new functor
+system has single APIs for both non-AD and AD property types.
+
+Currently, functor material property evaluations are defined using the API:
```
[the remark this answers is not in the document]
Comment (lindsayad) [reply framework/doc/content/syntax/FunctorMaterials/index.md:62]: It hasn't changed for a while (although it did change once before), so probably not

File: framework/doc/content/syntax/Functors/index.md [hunk 30]
```diff
@@ -0,0 +1,138 @@
...
+
+where here `_foo` is a functor data member of the kernel, `makeElemArg` is a helper routine for creating a
+functor element-based spatial argument, and `determineState()` is a helper routine for determining the correct
+time state to evaluate at, e.g. the current time for an implicit kernel and the old time for an explicit kernel.
```
Comment (lindsayad) [inline framework/doc/content/syntax/Functors/index.md:37]: I believe this is all in the `FunctorMaterials/index.md` page as well. I think some duplication is fine, although this seems like a lot

File: framework/doc/content/syntax/Functors/index.md [hunk 31]
```diff
@@ -0,0 +1,138 @@
...
+create specific implementations for specific arguments (as illustrated in `IMakeMyOwnFunctorProps`
+test class) or simply wishes to know more about the system, we give the details below.
+
+Any call to a functor looks like the following
+`_foo(const SpatialArg & r, const TemporalArg & t)`. Below are the possible type overloads of
+`SpatialArg`.
+
+#### FaceArg
+
+A typedef defining a "face" evaluation calling argument. This is composed of
```
Comment (lindsayad) [inline framework/doc/content/syntax/Functors/index.md:55]:
```suggestion
A `struct` defining a "face" evaluation calling argument. This is composed of
```
If we want to be C++ technical

File: framework/doc/content/syntax/Functors/index.md [hunk 32]
```diff
@@ -0,0 +1,138 @@
...
+- a face information object which defines our location in space
+- a limiter which defines how the functor evaluated on either side of the face should be
+  interpolated to the face
+- a boolean which states whether the face information element is upwind of the face
+- a boolean to indicate whether to correct for element skewness
+- a pair of subdomain IDs. These do not always correspond to the face info element subdomain
+  ID and face info neighbor subdomain ID. For instance if a flux kernel is operating at a
+  subdomain boundary on which the kernel is defined on one side but not the other, the
+  passed-in subdomain IDs will both correspond to the subdomain ID that the flux kernel is
+  defined on
```
Comment (lindsayad) [inline framework/doc/content/syntax/Functors/index.md:66]: This has been replaced with a `const Elem * face_side` data member

File: framework/src/materials/MaterialFunctorConverter.C [hunk 33]
```diff
@@ -0,0 +1,106 @@
...
+{
+  const auto & functors_in = getParam<std::vector<MooseFunctorName>>("functors_in");
+  const auto & reg_props_out = getParam<std::vector<MaterialPropertyName>>("reg_props_out");
+  const auto & ad_props_out = getParam<std::vector<MaterialPropertyName>>("ad_props_out");
+
+  if (isParamValid("reg_props_out") && isParamValid("ad_props_out"))
+    paramError("reg_props_out",
+               "We dont support converting functors to both AD and regular material properties. "
+               "Please use another " +
+                   type() + " to convert to both types.");
```
Comment (lindsayad) [inline framework/src/materials/MaterialFunctorConverter.C:45]:
```suggestion
               "We dont support converting functors to both regular and AD material properties in a single instance of '", type(), "'. Please create two instances, one for regular and one for AD.");
```

**Review by lindsayad — CHANGES_REQUESTED** [review]

File: framework/doc/content/syntax/Functors/index.md (hunk 27 above)
Comment (GiudGiud) [reply framework/doc/content/syntax/Functors/index.md:33]: You're right. I need to move more docs from FunctorMaterials.

File: framework/doc/content/syntax/Functions/index.md [hunk 34]
```diff
@@ -37,6 +37,20 @@ Check out `PiecewiseBilinear` to see how to update a function to support AD by
 using a templated `valueInternal()` function with virtual `value()` forwarders.
 
 
+### Functions as Functors
+
+Functions are [Functors](syntax/Functors/index.md). Functors are an abstraction, a base class, for
+objects that can compute values at a location in space and time.
+
+As `Functors`, they may be specified to objects such as the
+[FunctorElementalAux.md] in their functor argument. This vastly expands the number
```
Comment (GiudGiud) [inline framework/doc/content/syntax/Functions/index.md:46]:
```suggestion
[FunctorElementalAux.md] in their [!param](/AuxKernels/FunctorElementalAux/functor) parameter. This vastly expands the number
```

File: framework/doc/content/syntax/FunctorMaterials/index.md [hunk 35]
```diff
@@ -0,0 +1,108 @@
+# FunctorMaterials
+
+Functor material properties are properties that are evaluated on-the-fly.
+Along with multiple other classes of MOOSE objects, they are [Functors](syntax/Functors/index.md). This allows
+for the creation of objects with considerable inter-operability between systems.
+
+## Using functor materials
+
+Much like regular materials declare regular material properties, functor materials declare functor material properties.
```
Comment (GiudGiud) [inline framework/doc/content/syntax/FunctorMaterials/index.md:9]:
```suggestion

Much like regular materials declare regular material properties, functor materials declare functor material properties.
Functor material properties are properties that are evaluated on-the-fly.
Along with multiple other classes of MOOSE objects, they are [Functors](syntax/Functors/index.md). This allows
for the creation of objects with considerable inter-operability between systems.

## Using functor materials

```

File: framework/doc/content/syntax/FunctorMaterials/index.md [hunk 36]
```diff
@@ -0,0 +1,108 @@
...
+Functor materials are generally **NOT** compatible with regular materials, as regular material properties **CANNOT** be used
+in lieu of functor material properties. Functor material properties can only be used as regular material properties using a
+[MaterialFunctorConverter.md] to process them.
+
+Functor materials can be created within the `[FunctorMaterials]` block, and currently the `[Materials]` block as well,
+but that syntax will be deprecated.
+
+!alert note
+If a Functor is reported as missing by the simulation, and it was supposed to be created by a `FunctorMaterial`,
+you may use the [!param](/Debug/SetupDebugAction/show_functors) to get more information about what functors were
```
Comment (GiudGiud) [inline framework/doc/content/syntax/FunctorMaterials/index.md:29]:
```suggestion
you may use the [!param](/Debug/SetupDebugAction/show_functors) parameter to get more information about what functors were
```

File: framework/doc/content/syntax/FunctorMaterials/index.md (hunk 29 above)
Comment (GiudGiud) [inline framework/doc/content/syntax/FunctorMaterials/index.md:62]:
```suggestion
Functor material property evaluations are defined using the API:
```

File: framework/doc/content/syntax/Functors/index.md [hunk 37]
```diff
@@ -0,0 +1,138 @@
+# Functor system
+
+Functors are an abstraction, existing as a base class, that is available to several systems in MOOSE:
```
[the remark this answers is not in the document]
Comment (GiudGiud) [reply framework/doc/content/syntax/Functors/index.md:3]: It's more that they can be used as functors.
It s more an added capability than a restriction. Maybe there's a better word to express that

File: framework/doc/content/syntax/Functors/index.md (hunk 30 above)
Comment (GiudGiud) [reply framework/doc/content/syntax/Functors/index.md:37]: I ll go back and check

File: framework/src/materials/MaterialFunctorConverter.C (hunk 33 above)
Comment (GiudGiud) [reply framework/src/materials/MaterialFunctorConverter.C]: done

File: framework/doc/content/syntax/Functors/index.md (hunk 32 above)
Comment (GiudGiud) [reply framework/doc/content/syntax/Functors/index.md]: changed

File: framework/doc/content/syntax/Functors/index.md (hunk 30 above)
Comment (GiudGiud) [reply framework/doc/content/syntax/Functors/index.md:71]: made a link and more subsections

File: framework/doc/content/syntax/FunctorMaterials/index.md (hunk 29 above)
[the remark this answers is not in the document]
Comment (GiudGiud) [reply framework/doc/content/syntax/FunctorMaterials/index.md]: Removed current

File: framework/doc/content/syntax/FunctorMaterials/index.md (hunk 36 above)
[the remark this answers is not in the document]
Comment (GiudGiud) [reply framework/doc/content/syntax/FunctorMaterials/index.md]: added parameter after

File: framework/doc/content/syntax/FunctorMaterials/index.md (hunk 35 above)
[the remark this answers is not in the document]
Comment (GiudGiud) [reply framework/doc/content/syntax/FunctorMaterials/index.md]: done

File: framework/doc/content/syntax/FunctorMaterials/index.md [hunk 38]
```diff
@@ -0,0 +1,108 @@
...
+functor materials, etc, may consume.
+Both regular and functor material properties support a wide array of data types (scalar, vector, tensors,
+[automatic differentiation](automatic_differentiation/index.md) versions of all of these). Functor materials, because they are evaluated on-the-fly
+and call other functors on-the-fly as well, do not require dependency ordering.
+
+All functor materials support caching in some capacity. This can avoid expensive material property computations, but is
+disabled by default due to the potential memory cost. The reader is referred to the [Functors caching documentation](syntax/Functors/index.md#caching)
+
+!alert warning
+Functor materials are generally **NOT** compatible with regular materials, as regular material properties **CANNOT** be used
```
[the remark this answers is not in the document]
Comment (GiudGiud) [reply framework/doc/content/syntax/FunctorMaterials/index.md]: reworded

**Review by lindsayad — DISMISSED** [review]
great stuff

File: framework/doc/content/syntax/FunctorMaterials/index.md (hunk 36 above)
Comment (joshuahansel) [reply framework/doc/content/syntax/FunctorMaterials/index.md]: What I meant was that there's no indication that this is a parameter that needs to be in a `Debug` block.

File: framework/doc/content/syntax/Functions/index.md (hunk 28 above)
Comment (joshuahansel) [reply framework/doc/content/syntax/Functions/index.md:52]: Yeah, it's mainly something we need to communicate to reviewers.

File: framework/doc/content/syntax/FunctorMaterials/index.md (hunk 36 above)
[the remark this answers is not in the document]
Comment (joshuahansel) [reply framework/doc/content/syntax/FunctorMaterials/index.md]: I put this comment on some outdated suggestion I think: the problem is more that there's no indication that this "show_functors" parameter is a `Debug` block parameter.

[the remark this answers is not in the document]
Comment (GiudGiud) [reply framework/doc/content/syntax/FunctorMaterials/index.md]: will change it

**Comment by joshuahansel** [comment]
Is the conversion utility news-worthy?

**Comment by GiudGiud** [comment]
I ll add it to Casey's PR directly. I will hopefully get two more PRs in today so I can do them all at once

**Comment by GiudGiud** [comment]
please let me know when you're good with this review, I ll push all the changes together

**Comment by joshuahansel** [comment]
> please let me know when you're good with this review, I ll push all the changes together

I'll have more changes that I'd like, but nothing here is inaccurate - just potential improvements, so I'll approve after you implement the requested changes, and then at some point I'll probably make my own functor doc improvement PR after I get my feet wet with them and understand them more.

**Comment by GiudGiud** [comment]
Improved documentation after your experience sounds ideal!

**Review by joshuahansel — APPROVED** [review]

**Pull request merged**